flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/snapshot/StateSnapshotObserver.java

```diff
@@ -0,0 +1,197 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.operator.observer.snapshot;
+
+import org.apache.flink.autoscaler.utils.DateTimeUtils;
+import org.apache.flink.kubernetes.operator.api.FlinkDeployment;
+import org.apache.flink.kubernetes.operator.api.FlinkStateSnapshot;
+import org.apache.flink.kubernetes.operator.api.status.FlinkStateSnapshotState;
+import org.apache.flink.kubernetes.operator.controller.FlinkResourceContext;
+import org.apache.flink.kubernetes.operator.controller.FlinkStateSnapshotContext;
+import org.apache.flink.kubernetes.operator.observer.CheckpointFetchResult;
+import org.apache.flink.kubernetes.operator.observer.SavepointFetchResult;
+import org.apache.flink.kubernetes.operator.reconciler.ReconciliationUtils;
+import org.apache.flink.kubernetes.operator.service.FlinkResourceContextFactory;
+import org.apache.flink.kubernetes.operator.utils.EventRecorder;
+
+import io.fabric8.kubernetes.client.KubernetesClient;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+
+/** The observer of {@link org.apache.flink.kubernetes.operator.api.FlinkStateSnapshot}. */
+@RequiredArgsConstructor
+public class StateSnapshotObserver {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StateSnapshotObserver.class);
+
+    private final FlinkResourceContextFactory ctxFactory;
+    private final EventRecorder eventRecorder;
+
+    public void observe(FlinkStateSnapshotContext ctx) {
+        var resource = ctx.getResource();
+        var savepointState = resource.getStatus().getState();
+
+        if (FlinkStateSnapshotState.IN_PROGRESS.equals(savepointState)) {
+            observeSavepointState(ctx);
+        }
+    }
+
+    private void observeSavepointState(FlinkStateSnapshotContext ctx) {
+        var resource = ctx.getResource();
+        var resourceName = resource.getMetadata().getName();
+        var triggerId = resource.getStatus().getTriggerId();
+
+        if (StringUtils.isEmpty(triggerId)) {
+            return;
+        }
+
+        LOG.debug("Observing savepoint state for resource {}...", resourceName);
+        var secondaryResource =
+                ctx.getSecondaryResource()
+                        .orElseThrow(
+                                () ->
+                                        new RuntimeException(
+                                                String.format(
+                                                        "Secondary resource %s for savepoint %s was not found",
+                                                        resource.getSpec().getJobReference(),
+                                                        resourceName)));
```

Comment:
If the target is not found anymore to me that sounds similar to the not running case, and probably should be marked abandoned. What is your thinking here?

Reply:
Yeah, I agree this should be abandoned as well. I have also added an event that will be produced when the snapshot gets to an abandoned state.